testonly/hammer/hammer.go

```diff
@@ -181,15 +185,55 @@ func HitMap(cfg MapConfig) error {
 		}
 	}(ticker.C)
 
-	for count := uint64(1); count < cfg.Operations; count++ {
-		if err := s.retryOneOp(ctx); err != nil {
-			return err
+	var wg sync.WaitGroup
+	errs := make(chan error, cfg.NumCheckers+1)
+	done := make(chan struct{})
+	for i := 0; i < cfg.NumCheckers; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			glog.Infof("%d: start checker %d", s.cfg.MapID, i)
+			err := s.readChecker(ctx, done, i)
+			if err != nil {
+				errs <- err
+			}
+			glog.Infof("%d: checker %d done with %v", s.cfg.MapID, i, err)
+		}(i)
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		glog.Infof("%d: start main goroutine", s.cfg.MapID)
+		count, err := s.performOperations(ctx, done)
+		errs <- err // may be nil for the main goroutine completion
+		glog.Infof("%d: performed %d operations on map", cfg.MapID, count)
+	}()
+
+	// Wait for first error, completion (which shows up as a nil error) or
+	// external cancellation.
+	var firstErr error
+	select {
+	case <-ctx.Done():
+		glog.Infof("%d: context canceled", cfg.MapID)
+	case e := <-errs:
+		firstErr = e
+		if firstErr != nil {
+			glog.Infof("%d: first error encountered: %v", cfg.MapID, e)
 		}
 	}
-	glog.Infof("%d: completed %d operations on map", cfg.MapID, cfg.Operations)
-	ticker.Stop()
+	close(done)
 
-	return nil
+	ticker.Stop()
+	wg.Wait()
+	close(errs)
+	for e := range errs {
+		if e == nil {
```

Comment:
Go style: The default code path should be the one with the least indentation

Reply:
I originally had it the other way round and it looked odd -- for iterating through errors, `e == nil` is kind-of the non-default path (there can only ever be one `nil`, but there could be N non-`nil` errors).